### PR TITLE
Improve syncing

### PR DESCRIPTION
**Pull request by joefitter**
* only allow one sync at a time, keep syncing until everything up to date
* added unbeforeunload if syncing
* disabled next steps if syncing

### CODE REVIEW

File: client/project-router.js [hunk 1]
```diff
@@ -8,11 +9,36 @@ import Project from './pages/project';
...
+  const isSyncing = useSelector(state => state.application.isSyncing);
 
   function toggleStatusShowing() {
     setStatusShowing(!statusShowing);
   }
 
+  useEffect(() => {
+    window.onbeforeunload = () => {
+      if (isSyncing) {
+        return 'Data is currently syncing, if you leave this page your recent changes may not be saved.';
```
Comment (lennym) [inline client/project-router.js]:
```suggestion
        return 'Data is currently being saved, if you leave this page your recent changes may not be saved.';
```

**Comment by lennym** [comment]
It's not quite there. With this code as is, then if the conditions are being synced (and so `inSyncing` is true) then a change to `retrospectiveAssessment` is completely disregarded.

I think the check for changed fields need to incorporate all possible fields, and sync _any_ changes if found, rather than just syncing one field at a time.

An alternative approach would be to remove the condition, but move the "is there a sync in progress" condition inside `sync` and add a delay and recursive call in an `else` clause.

File: client/project-router.js [hunk 2]
```diff
@@ -8,11 +9,40 @@ import Project from './pages/project';
...
+      }
+    }
+
+    const nextSteps = document.querySelector('#page-component > p.next-steps > a');
+    const statusMessage = document.querySelector('#page-component > p.next-steps > span.status-message');
+
+    if (isSyncing) {
+      nextSteps.setAttribute('disabled', true);
+      nextSteps.onclick = () => false;
+      statusMessage.innerText = 'Saving...';
```
Comment (lennym) [inline client/project-router.js]:
```suggestion
      statusMessage && statusMessage.innerText = 'Saving...';
```

File: client/project-router.js [hunk 3]
```diff
@@ -8,11 +9,40 @@ import Project from './pages/project';
...
+    } else {
+      nextSteps.removeAttribute('disabled');
+      nextSteps.onclick = null;
+      statusMessage.innerText = '';
```
Comment (lennym) [inline client/project-router.js]:
```suggestion
      statusMessage && statusMessage.innerText = '';
```

File: client/project-router.js [hunk 4]
```diff
@@ -8,11 +9,40 @@ import Project from './pages/project';
...
+    }
+
+    return () => {
+      window.onbeforeunload = null;
+      nextSteps.removeAttribute('disabled');
+      nextSteps.onclick = null;
+      statusMessage.innerText = '';
```
Comment (lennym) [inline client/project-router.js]:
```suggestion
      statusMessage && statusMessage.innerText = '';
```

File: client/actions/projects.js [hunk 5]
```diff
@@ -106,20 +119,47 @@ export function updateAndSave(data) {
...
-      dispatch(throwError('Error updating conditions'));
+      dispatch(doneSyncing());
+      return sync(dispatch, getState, fields, extra);
     });
-}
+};
+
+const debouncedSync = debounce((dispatch, getState, extra) => {
+  return sync(dispatch, getState, ['conditions', 'retrospectiveAssessment'], extra)
+}, 1000, { maxWait: 5000 });
```
Comment (wheelsandcogs) [inline client/actions/projects.js]: Does this mean we do a max of 1 sync per second, with a 5 second timeout?

Comment (joefitter) [reply client/actions/projects.js]: it means it waits for a 1 second pause before trying to sync again, up to a max of 5 seconds. Once the sync is called it starts the recursive dance

**Pull request merged**